pyrules/conditions.py

```diff
@@ -0,0 +1,238 @@
+
+
+class C(object):
+    AND = 'AND'
+    OR = 'OR'
+    XOR = 'XOR'
+    default = AND
+    OPS_MAP = {AND: operator.and_, OR: operator.or_, XOR: operator.xor}
+
+    def __init__(self, _key=None, _value=None, _negate=False, **kwargs):
+        if _key:
+            self._expr = (_key, _value, _negate)
+        else:
+            self._expr = ()
+        for k, v in kwargs.iteritems():
+            self._expr = self.combine(C(k, v))._expr
+
+    def clone(self):
+        """
+        Create a new C-expression that is a copy of this one.
+        """
+        new = self.__class__()
+        new._expr = self._expr
+        return new
+            
+    def combine(self, other, op=default):
+        """
+        Combine another C-expression with this using a logical operator.
+        """
+        assert(isinstance(other, C))
+        self._expr = (
+            op, self._expr, other._expr, False) if self._expr else other._expr
+        return self
+
+    def apply(self, context):
+        """
+        Apply evaluation context to this C-expression. Context should be a dict
+        or dict-like object.
+        """
+        return self._apply_expr(context, self._expr)
+
+    def _apply_expr(self, context, expr):
+        if isinstance(expr[1], tuple):
+            return self.OPS_MAP[expr[0]](
+                self._apply_expr(context, expr[1]),
+                self._apply_expr(context, expr[2]))
+        else:
+            # expr[2] holds negation flag, that's why we xor with it.
+            return ExpressionHandler().evaluate(
+                expr[0], context, expr[1]) ^ expr[2]
+
+    def __repr__(self):
+        return '<C: {}>'.format(self._to_str(self._expr)).encode('utf-8')
+
+    def _to_str(self, val):
+        if not val:
+            return u'N/A'
+        elif isinstance(val[1], tuple):
+            return u'{}({} {} {})'.format(
+                'NOT' if val[-1] else '', self._to_str(val[1]), val[0],
+                self._to_str(val[2]))
+        else:
+            return u'{} {}= {}'.format(
+                val[0], '!' if val[-1] else '=', val[1])            
+
+    def __and__(self, other):
+        return self.clone().combine(other, self.AND)
+
+    def __or__(self, other):
+        return self.clone().combine(other, self.OR)
+
+    def __xor__(self, other):
+        return self.clone().combine(other, self.XOR)
+
+    def __invert__(self):
+        new = self.clone()
+        new._expr = tuple(self._expr[:-1]) + (not self._expr[-1],)
+        return new
+
+    def __eq__(self, other):
+        if not issubclass(other.__class__, C):
+            return False
+        return self._expr == other._expr
+
+
+class ExpressionError(Exception):
+    pass
+
+        
+class ExpressionHandler(object):
+    handlers = [
+        'gt', 'lt', 'gte', 'lte', 'eq', 'neq', 'contains', 'icontains', 'bool']
+
+    def evaluate(self, expression, context, arg):
+        bits = expression.split('__')
```

Comment:
What's the rationale for doing parsing at this point?

I'd prefer to have parsing done by the C object rather than by the ExpressionHandler - as it is now, we will never know whether expressions work until the condition is actually executed. I.e. the C object should parse the k/v pairs into key, op and value:

```
C(foo__eq=5)
=> key=foo, op=eq, value=5
```

This way, if the operator is not given it defaults to `eq`, if it is unknown we can raise an exception at parsing time.